### PR TITLE
Refactor in preparation for arbitrary grid development

### PR DESCRIPTION
**Pull request by kubagalecki**
- Factored out elements which will be common between cartesian and arbitrary grids into `LatticeData`
- Factored out `Node`
- Introduced `[Cart]LatticeExecutor` as an intermediate abstraction for running kernels
- Stripped down `[Cart]LatticeContainer` to functionality related to storage
- Linearized dependencies between `[Cart]LatticeAccess` and `[Cart]LatticeContainer`
- General cleanup of `cuda.cu`

### CODE REVIEW

File: src/Lattice.h.Rt [hunk 1]
```diff
@@ -39,7 +39,7 @@ const int maxSnaps=33;
 class Lattice {
 private:
   Model_m my_model;
-  LatticeContainer * container; ///< Main Container object
+  CartLatticeExecutor* executor; ///< Main execution context, including the lattice container, data, etc.
```
Comment (llaniewski) [inline src/Lattice.h.Rt:42]: Change from pointer to object.

File: src/Global.h.Rt [hunk 2]
```diff
@@ -169,6 +169,61 @@ int kbhit(void);
 #define DEBUG_H 1
 #endif
 
+/// Allocation of memory for an FTabs
+void FTabs::Alloc(int nx,int ny,int nz) {
```
Comment (llaniewski) [inline src/Global.h.Rt:173]: I think this is not used

File: src/Global.h.Rt [hunk 3]
```diff
@@ -65,10 +65,10 @@ for (m in Margin) { ?>
     for (m in NonEmptyMargin) { ?>
```
Comment (llaniewski) [inline src/Global.h.Rt:65]: I think `FTabs` should be moved from here. Probably to `CartLatticeContainer.h`

File: src/CartLatticeExecutor.h.Rt [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+  LatticeData          m_data;
+
+public:
+  template < eOperationType I, eCalculateGlobals G, eStage S >
+  void RunBorder(CudaStream_t stream) const;
+  template < eOperationType I, eCalculateGlobals G, eStage S >
+  void RunInterior(CudaStream_t stream) const;
+  void Color(uchar4 *out_ptr);
+
+  CartLatticeContainer&       getContainer()       { return m_container; }
```
Comment (llaniewski) [inline src/CartLatticeExecutor.h.Rt:25]: I woulde remove the getters, and the `m_` (as it's not used in the code elsewhere).

File: src/Solver.cpp.Rt [hunk 5]
```diff
@@ -12,8 +12,8 @@
 #endif
 #include "cross.h"
 #include "Region.h"
-//#include "LatticeContainer.h"
-class LatticeContainer;
+//#include "CartLatticeContainer.h"
+class CartLatticeContainer;
```
Comment (llaniewski) [inline src/Solver.cpp.Rt:16]: I think this is obsolete

File: src/CartLatticeExecutor.hpp.Rt [hunk 6]
```diff
@@ -0,0 +1,337 @@
...
+//  }
+  tab[r.offset(x,y,z)] = w;
+}
+
+<?R
+  }
+  ifdef();
+?>
+
+<?R   for (tp in rows(AllKernels)[order(AllKernels$adjoint)]) {
```
Comment (llaniewski) [inline src/CartLatticeExecutor.hpp.Rt:303]: I would like to have this (template instantiations) in 'cuda.cu'

**Review by llaniewski — CHANGES_REQUESTED** [review]

File: src/CartLatticeExecutor.hpp.Rt (hunk 6 above)
Comment (kubagalecki) [reply src/CartLatticeExecutor.hpp.Rt]: Closed in [375c34e](https://github.com/CFD-GO/TCLB/pull/454/commits/375c34e5f21feba76b06191ad6c07c2c20fa4da0)

File: src/Solver.cpp.Rt (hunk 5 above)
Comment (kubagalecki) [reply src/Solver.cpp.Rt]: Closed in [066f2e7](https://github.com/CFD-GO/TCLB/pull/454/commits/066f2e7f930cf11249a8fcb6006ce47972e0dcdc)

File: src/CartLatticeExecutor.h.Rt (hunk 4 above)
Comment (kubagalecki) [reply src/CartLatticeExecutor.h.Rt]: Closed in cd5e88a59be1e4ba72f543c06f6e35e1c8c91aad

File: src/Global.h.Rt (hunk 2 above)
Comment (kubagalecki) [reply src/Global.h.Rt]: Closed in 2c27d9263125e8726e2bb5c921b47491b6e4690c

File: src/CartLatticeContainer.h.Rt [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+  flag_t * NodeType; ///< Table of flags/NodeTypes of all the nodes
+  cut_t* Q; 
+  int nx, ny, nz; ///< Size of the Lattice region
+  real_t px,py,pz;
+
+  inline void Alloc (int,int,int);
+  inline void Free();
+  inline void ActivateCuts();
+  
+  CudaDeviceFunction flag_t getType(int x, int y, int z) const { return NodeType[(x + nx*(y + ny*z))]; }
```
[the remark this answers is not in the document]
Comment (kubagalecki) [reply src/CartLatticeContainer.h.Rt]: Closed in [01f978a](https://github.com/CFD-GO/TCLB/pull/454/commits/01f978a2a7460523bd8d72fd88ced25887aa001a)

File: src/Lattice.h.Rt (hunk 1 above)
Comment (kubagalecki) [reply src/Lattice.h.Rt]: Closed in b18a64d39c6de22061dc02e4c5ce95c649d98731

File: src/CartLatticeContainer.hpp.Rt [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+/** \file CartLatticeContainer.hpp
+*/
+
+#ifndef CARTLATTICECONTAINER_HPP
+#define CARTLATTICECONTAINER_HPP
+
+#include "CartLatticeContainer.h"
+
+/// Allocation of memory of a container
+void CartLatticeContainer::Alloc(int nx_, int ny_, int nz_)
```
[the remark this answers is not in the document]
Comment (kubagalecki) [reply src/CartLatticeContainer.hpp.Rt]: Closed in [3df758e](https://github.com/CFD-GO/TCLB/pull/454/commits/3df758e39ed7393283f40b1100474a0ba157826a)

File: src/Global.h.Rt (hunk 3 above)
Comment (kubagalecki) [reply src/Global.h.Rt]: Closed in [f47e01e](https://github.com/CFD-GO/TCLB/pull/454/commits/f47e01eca95590835f8efeb77e917ea56261d979)

Where should we put the definitions for FTabs though? Just inline them?

Comment (kubagalecki) [reply src/Global.h.Rt]: @llaniewski verbally confirmed they can stay in `Lattice.cu`

File: src/CartLatticeAccess.hpp.Rt [hunk 9]
```diff
@@ -253,18 +250,22 @@ resolve.symmetries = function(D) {
 
 ?>
 
+CudaDeviceFunction flag_t getContainerType(const CartLatticeContainer& container, int x, int y, int z) {
```
Comment (llaniewski) [inline src/CartLatticeAccess.hpp.Rt:253]: This name is misleading, as it suggest container type. Maybe `getNodeType`? And then we'll overload it for different lattices.

File: src/CartLatticeAccess.hpp.Rt [hunk 10]
```diff
@@ -253,18 +250,22 @@ resolve.symmetries = function(D) {
...
   const y_t y;
   const z_t z;
   const flag_t nt;
   const range_int<0,1,0,1> nx, ny, nz;
   public:
-  CudaDeviceFunction LatticeAccess(const int& x_, const int& y_, const int& z_) :
+  CudaDeviceFunction CartLatticeAccess(const int& x_, const int& y_, const int& z_) :
     x(x_),y(y_),z(z_),
-    nt(constContainer.getType(x,y,z)),
+    nt{getContainerType(constContainer, x, y, z)},
```
Comment (llaniewski) [inline src/CartLatticeAccess.hpp.Rt:268]: Can we use normal brackets in object initializations? There is not a lot of difference and let's just make it consistent with the rest of the code.

**Review by llaniewski — CHANGES_REQUESTED** [review]
Only cosmetic things.

**Review by llaniewski — APPROVED** [review]

**Pull request merged**